sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringOperations.scala

```diff
@@ -298,3 +299,56 @@ case class StringLength(child: Expression) extends UnaryExpression with ExpectsI
 
   override def prettyName: String = "length"
 }
+
+/**
+ *  A function that return soundex code of the given string expression.
+ */
+case class SoundEx(child: Expression) extends UnaryExpression with ExpectsInputTypes {
+  @transient lazy val so = new Soundex
+
+  override def dataType: DataType = StringType
+
+  override def inputTypes: Seq[DataType] = Seq(StringType)
+
+  override def eval(input: InternalRow): Any = {
+    val string = child.eval(input)
+    if (string == null) {
+      null
+    } else {
+      val str: String = string.toString
+      UTF8String.fromString(so.encode(str))
+    }
+  }
+}
+
+/**
+ * Returns string, with the first letter of each word in uppercase,
+ * all other letters in lowercase. Words are delimited by whitespace.
+ */
+case class InitCap(child: Expression) extends UnaryExpression with ExpectsInputTypes {
+  @transient lazy val so = new Soundex
+
+  override def dataType: DataType = StringType
+
+  override def inputTypes: Seq[DataType] = Seq(StringType)
+
+  override def eval(input: InternalRow): Any = {
+    val string = child.eval(input)
+    if (string == null) {
+      null
+    } else {
+      val str: String = string.toString
+      val sb = new StringBuilder
```

Comment:
On the surface this might be true, but look in the implementation
http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/lang/AbstractStringBuilder.java#AbstractStringBuilder.append%28java.lang.CharSequence%2Cint%2Cint%29

http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/java/lang/AbstractStringBuilder.java#AbstractStringBuilder.charAt%28int%29

`append` calls `charAt` `length` times. We would save every `chartAt` call that is triggered when you replace a character. The second optimisation is to initialise `StringBuilder` with `str.length`. This avoids that `StringBuilder` has to expand it's size.